File: Kotlin/ch04/src/main/kotlin/ch04/BinaryTree.kt
package ch04

/**
 * 이진 트리 (Binary Tree)
 * Rust/ch04/src/binary_tree.rs → Kotlin 이디엄 포팅
 */
class BinaryNode(val data: Char) {
    var left: BinaryNode? = null
    var right: BinaryNode? = null
}

fun preorder(node: BinaryNode?, out: MutableList<Char> = mutableListOf()): List<Char> {
    if (node != null) {
        out.add(node.data)
        preorder(node.left, out)
        preorder(node.right, out)
    }
    return out
}

fun inorder(node: BinaryNode?, out: MutableList<Char> = mutableListOf()): List<Char> {
    if (node != null) {
        inorder(node.left, out)
        out.add(node.data)
        inorder(node.right, out)
    }
    return out
}

fun postorder(node: BinaryNode?, out: MutableList<Char> = mutableListOf()): List<Char> {
    if (node != null) {
        postorder(node.left, out)
        postorder(node.right, out)
        out.add(node.data)
    }
    return out
}

fun main() {
    //        A
    //      /   \
    //     B     E
    //    / \   / \
    //   C   D F   G
    val b = BinaryNode('B').also {
        it.left = BinaryNode('C')
        it.right = BinaryNode('D')
    }
    val e = BinaryNode('E').also {
        it.left = BinaryNode('F')
        it.right = BinaryNode('G')
    }
    val a = BinaryNode('A').also {
        it.left = b
        it.right = e
    }

    println("Preorder:  ${preorder(a)}")
    println("Inorder:   ${inorder(a)}")
    println("Postorder: ${postorder(a)}")
}

File: Rust/ch04/src/bin/binary_tree.rs
// 이진 트리 (Binary Tree)
// Clang/04/BinaryTree 포팅. 로직은 ch04::binary_tree 모듈에 있다.
use ch04::binary_tree::{inorder_print_tree, postorder_print_tree, preorder_print_tree, SBTNode};

fn main() {
    // 노드 생성 및 트리 구성 (C 테스트와 동일한 구조)
    //        A
    //      /   \
    //     B     E
    //    / \   / \
    //   C   D F   G
    let mut b = SBTNode::new('B');
    b.left = Some(SBTNode::new('C'));
    b.right = Some(SBTNode::new('D'));

    let mut e = SBTNode::new('E');
    e.left = Some(SBTNode::new('F'));
    e.right = Some(SBTNode::new('G'));

    let mut a = SBTNode::new('A');
    a.left = Some(b);
    a.right = Some(e);

    let root = Some(a);

    // 트리 출력
    println!("Preorder ...");
    preorder_print_tree(&root);
    println!("\n");

    println!("Inorder ... ");
    inorder_print_tree(&root);
    println!("\n");

    println!("Postorder ... ");
    postorder_print_tree(&root);
    println!();
}
